### PR TITLE
[FEATURE] Fluent Datasource config substitution support

### PR DESCRIPTION
**Pull request by Kilo59**
## Changes proposed in this pull request:
- [x] substitute `${FOO}` variables set for `xdatasources`/ Fluent-Datasources when loading from yaml
- [x] retain `${FOO}` values when persisting back to config
   - [x] make sure this still works if assets/datasources are modified after loading from config
      - [x] fix the tests 
- [ ] ensure sensitive values do not leak as part of validation errors
- [ ] cleanup comments/debugging

### Definition of Done

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.

### CODE REVIEW

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 1]
```diff
@@ -5420,11 +5420,11 @@ def save_config_variable(
             yaml.dump(config_variables, config_variables_file)
 
     def _load_zep_config(self) -> GxConfig:
-        """Called at beginning of DataContext __init__"""
+        """Called at beginning of DataContext __init__ after config_providers init."""
         logger.info(
             f"{self.__class__.__name__} has not implemented `_load_zep_config()` returning empty `GxConfig`"
         )
-        return GxConfig(xdatasources={})
+        return GxConfig(xdatasources={}, _config_provider=self.config_provider)
```
Comment (Kilo59) [inline great_expectations/data_context/data_context/abstract_data_context.py]:
```suggestion
        return GxConfig(xdatasources={})
```

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 2]
```diff
@@ -268,6 +266,8 @@ def __init__(self, runtime_environment: Optional[dict] = None) -> None:
         self._config_variables = self._load_config_variables()
         self._variables = self._init_variables()
 
+        self.zep_config = self._load_zep_config()
```
Comment (Kilo59) [inline great_expectations/data_context/data_context/abstract_data_context.py:269]:
```suggestion
        # config providers must be provisioned before loading zep_config
        self.zep_config = self._load_zep_config()
```

File: great_expectations/experimental/datasources/config.py [hunk 3]
```diff
@@ -121,8 +147,23 @@ def parse_yaml(
...
+    def _load_yaml_with_config_substitutions(
+        cls: Type[GxConfig],
+        f: Union[pathlib.Path, str],
+        _config_provider: _ConfigurationProvider,
+    ) -> _ConfigsTuple:
+        raw_config = yaml.load(f)
+        logger.debug(f"Substituting config with {_config_provider.__class__.__name__}")
+        return _ConfigsTuple(
+            _config_provider.substitute_config(raw_config), raw=raw_config
+        )
```
Comment (Kilo59) [inline great_expectations/experimental/datasources/config.py]: This is probably not worth it's own method

File: great_expectations/data_context/data_context/serializable_data_context.py [hunk 4]
```diff
@@ -101,6 +103,18 @@ def _save_project_config(self) -> None:
                         f"Saving {len(self.zep_config.datasources)} ZEP Datasources to {config_filepath}"
                     )
                     zep_json_dict: dict[str, JSONValues] = self.zep_config._json_dict()
+                    # re-combine the original raw config to preserve the original config ${FOO} placeholders
+                    # NOTE: this could cause issues depending on what values have the placeholders
+                    # TODO: Only update fields that had config placeholders?
+                    # TODO: cleanup
+                    updated_xdatasources = deep_update(
```
Comment (billdirks) [inline great_expectations/data_context/data_context/serializable_data_context.py:110]: If a user deletes parts of the config could this deep_update restore them accidentally? I can't seem to find the documentation for it.

**Review by billdirks — COMMENTED** [review]
This isn't that much code which is great. My main concern is with resolving the before and after states of the config. That is, when a user loads a config, makes changes, and then serializes it back out. Maybe that's because I don't know how the `deep_update` works. If the user does lots of additions/deletions to the config, could this be a problem? Maybe it would be better to keep a mapping of the resolved_strings -> original_variable strings. Then, you could go through the final config and specifically substitute those strings if you find them.

File: great_expectations/data_context/data_context/serializable_data_context.py [hunk 5]
```diff
@@ -101,6 +103,18 @@ def _save_project_config(self) -> None:
...
+                        zep_json_dict["xdatasources"],  # type: ignore[arg-type]
+                        self.zep_config._xdatasources_raw_config,
+                    )
+                    zep_json_dict["xdatasources"] = updated_xdatasources
+                    # print(
+                    #     f"original\n{pf(zep_json_dict['xdatasources'])}\n\n\nupdated\n{pf(updated_xdatasources)}"
+                    # )
```
Comment (NathanFarmer) [inline great_expectations/data_context/data_context/serializable_data_context.py:117]: Is this supposed to be commented out?

File: great_expectations/data_context/data_context/serializable_data_context.py (hunk 4 above)
Comment (Kilo59) [reply great_expectations/data_context/data_context/serializable_data_context.py:110]: That's a great question. I assume I'll have to account for that.

File: great_expectations/data_context/data_context/serializable_data_context.py [hunk 6]
```diff
@@ -7,6 +7,8 @@
 import warnings
 from typing import TYPE_CHECKING, ClassVar, Optional, Union
 
+# from pprint import pformat as pf
+from pydantic.utils import deep_update
```
Comment (NathanFarmer) [inline great_expectations/data_context/data_context/serializable_data_context.py:11]: Does this update types? If not why is it a pydantic util?

File: great_expectations/data_context/data_context/serializable_data_context.py (hunk 5 above)
Comment (Kilo59) [reply great_expectations/data_context/data_context/serializable_data_context.py:117]: Will remove this before merge.

**Comment by Kilo59** [comment]
Closing this in-favor of a different approach.

**Pull request closed without merging**